File: L6/App_6_2/App_6_2.f90
MODULE L5
    IMPLICIT NONE

    CONTAINS

    SUBROUTINE resize(v,  old_size, new_size)
        INTEGER, ALLOCATABLE :: v(:)
        INTEGER :: new_size, old_size, i
        INTEGER, ALLOCATABLE :: t_ptr(:)

        ALLOCATE(t_ptr(new_size))

        DO i = 1, old_size
            t_ptr(i) = v(i)
        END DO

        DEALLOCATE(v)
        ALLOCATE(v(new_size))

        v = t_ptr

        DEALLOCATE(t_ptr)
    END SUBROUTINE resize

    SUBROUTINE insertion_sort(v, n)
        INTEGER, ALLOCATABLE :: v(:)
        INTEGER :: n, i, j, key

        DO i = 2, n
            key = v(i)
            j = i - 1

            DO WHILE(j >= 1 .AND. v(j) > key)
                v(j + 1) = v(j)
                j = j - 1
            END DO

            v(j + 1) = key
        END DO
    END SUBROUTINE insertion_sort
END MODULE L5

PROGRAM main
    USE L5
    IMPLICIT NONE

    INTEGER :: n, cap, x, stat, i
    INTEGER, ALLOCATABLE :: v(:)

    n = 0
    cap = 0

    OPEN(1, FILE="file.in", ACTION="read")

    ALLOCATE(v(0))

    DO
        READ(1, *, IOSTAT=stat) x
        
        IF (stat .NE. 0) EXIT

        IF (n >= cap) THEN
            cap = n * 2 + 1
            CALL resize(v, n, cap)
        END IF

        v(n + 1) = x
        n = n + 1
    END DO

    CALL insertion_sort(v, n)

    DO i = 1, n
        WRITE(*,*) v(i)
    END DO

    DEALLOCATE(v)
    CLOSE(1)
END PROGRAM main

File: L6/App_6_2/App_6_2.cpp
#include <iostream>
#include <fstream>

namespace L5
{
    void resize(int *&ptr, unsigned int old_size, unsigned int new_size)
    {
        if (new_size == 0)
            return;

        int *t_ptr = static_cast<int*>(malloc(new_size * sizeof(int)));

        if (t_ptr == nullptr)
            throw std::bad_alloc();

        for (unsigned int i = 0; i < old_size; ++i)
            t_ptr[i] = ptr[i];

        free(ptr);
        ptr = t_ptr;
    }
}

void insertion_sort(int *&v, unsigned int n)
{
    if (v == nullptr || n == 0)
        return;

    for (unsigned int i = 1; i < n; ++i)
    {
        int key = v[i], j = i - 1;

        while (j >= 0 && v[j] > key)
        {
            v[j + 1] = v[j];
            --j;
        }

        v[j + 1] = key;
    }
}

int main()
{
    std::ifstream fin("file.in");

    if (!fin.is_open())
    {
        std::cerr << "Failed to open file!\n";
        return -1;
    }

    unsigned int n = 0, cap = 0;
    int *v = nullptr, x;

    while (fin >> x)
    {
        if (n >= cap)
        {
            unsigned int new_cap = n * 2 + 1;
            L5::resize(v, cap, new_cap);
            cap = new_cap;
        }

        v[n] = x;
        ++n;
    }

    fin.close();

    insertion_sort(v, n);

    for (unsigned int i = 0; i < n; ++i)
        std::cout << v[i] << " ";

    std::ofstream fout("file.out");

    if (!fout.is_open())
    {
        std::cerr << "Failed to open file!\n";
        return -1;
    }

    for (unsigned int i = 0; i < n; ++i)
        fout << v[i] << " ";

    fout.close();

    free(v);

    return 0;
}
